src/core/qgsclipper.cpp

```diff
@@ -38,9 +37,68 @@ const double QgsClipper::MIN_Y = -16000;
 
 const double QgsClipper::SMALL_NUM = 1e-12;
 
-QPolygonF QgsClipper::clippedLine( const QgsCurve &curve, const QgsRectangle &clipExtent )
+QgsLineString QgsClipper::clipped3dLine( const QgsCurve &curve, const QgsBox3d &clipExtent )
 {
-  return clippedLine( curve.asQPolygonF(), clipExtent );
+  const int nPoints = curve.numPoints();
+
+  double p0x, p0y, p0z, p1x = 0.0, p1y = 0.0, p1z = 0.0; //original coordinates
+  double p1x_c, p1y_c, p1z_c; //clipped end coordinates
+  double lastClipX = 0.0, lastClipY = 0., lastClipZ = 0.0; //last successfully clipped coords
+
+  QgsLineString line;
+
+  QgsPoint curPoint;
+  QgsVertexId::VertexType type;
+  for ( int i = 0; i < nPoints; ++i )
```

Comment:
Maybe, you can use vertex_iterators here

Reply:
done!